utils/docs/sw2asc.1.in
.TH sw2asc 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
.B sw2asc
\- converts a SWITCAP2 output file into ASCII data files that other tools can read

.SH SYNOPSIS
.B sw2asc \-c\||\-\-copyright

.B sw2asc \-h\||\-\-help

.B sw2asc \-V\||\-\-version

.B sw2asc
.I switcap.out

.SH DESCRIPTION

The
.I sw2asc
program reads the switcap.out output file produced by SWITCAP2 and produces
a set of ASCII datafiles which can be more easily loaded into various waveform
plotting programs.

Data generated by SWITCAP2 "PRINT" statements is;
extracted and written to an output file in an ASCII format which can be
directly loaded by the
.I gwave
waveform viewer or more easily loaded into
a tool such as
.I scilab
,
.I octave
,
.I grace,
, or
.I matlab
.

If multiple analyses are in the SWITCAP2 output file, then they are written
to individual files.  The output file names are derived from the input
file name by appending ".{SSS,TRAN}.#.asc"" to the name of the input file.
SSS is used for Sinusoidal Steady State analysis output and TRAN is used for.
TRANsient analysis output.  # corresponds to the analysis number (ie if you
have 3 SSS analyses in your file, you will end up with .SSS.1.asc, .SSS.2.asc,
and .SSS.3.asc.

.SH OPTIONS
.TP
\-c, \-\-copyright
prints copyright information and exits
.TP
\-h, \-\-help
prints help message and exits
.TP
\-V, \-\-version
prints version information and exits

.SH EXIT STATUS
.I sw2asc
will return 0 on success and non-zero
if the command did not complete successfully.

.SH SEE ALSO
SWITCAP2 users manual
.PP
.SH COPYRIGHT
.nf
Copyright \(co 2003-@YEAR@ gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.



NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2015 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 2.1.0.0 - 04/01/15 Rampage Edition

* This release of the Rampage Edition of the gEDA-gaf suite commemorates
  the 17th Anniversary of the first public release of gEDA-gaf. The 2.09
  series releases reflected a very aggressive development of new features
  and major enhancements in many areas. Numerous enhancements focused on
  existing features and functions that did not always work as expected.

* New editing functions provide powerful methods to improve productivity.

  A new Extend Objects functions allows linear objects to be extended to
  other objects with a simple key press.

  A new Break function allows Arcs, Boxes, Circles, Linear objects, and
  Paths, both closed and open, to be broken at user specified points.

  A Powerful new Array dialog has been added to create arrays of selected
  objects. Creating Arrays allows both verb-noun or noun-verb operations.

  A new Auto-Snap Objects to Grid editing function has been added allowing
  Selections to be quickly repositioned to the current Grid snap. The Auto
  Snap features works with all objects types, including graphical and image
  objects.

  A new bimodal Offset function provides editing methods to quickly create
  new objects at programmable distances. The distance can be specified by
  entering a value in a dialog or offsetting Selection sets to the current
  pointer position using hot-keys. The offset distance is retained between
  operations.

  A new Select Last function has been added to select the last object created
  or edited.

  The visibility of netname attributes can now be toggled while inside an
  action.

  A new RC keyword; third-button-cancel, allows user to specify a preference
  for canceling operations using the third mouse button. In addition, object
  adder modes can now be cancel by double-clicking the final entry point.

  A new Zoom to Magnification feature allows user to zoom to a specified level.

* Picture replacement operations can now be undone.

* A new Export functions allow exporting of embedded objects, including image
  objects.

* Picture objects no longer need to maintain the original aspect ratio. Aspect
  ratios of image objects can be freely modified using Grips. The Ctrl key can
  be combined to maintain the aspect ratio identically to the old behavior.
  When exporting images using the new export feature, options are provided to
  export the image at the original or the current aspect ratio and size.

* Coordinates can now be displayed in the status bar using Cartesian and vector
  formats. The format can be changed at any time, include inside of an action,
  by right clicking on the coordinates widget and selecting the format from a
  pop-up menu. Vector formats revert to the previous selected format when not
  applicable to the active context.

* Pointer button assignments can now be set via a pop-up menu by clicking on
  the associated status widget on the Status bar. Assignments can still be set
  using in the Preferences Dialog or over-ridden using RC keywords, but the
  new method allows changing assignments while inside an action.

* The status text now changes to blue when inside of an editing action.

* F-Keys have been assigned to common/standard program functions.

* Menus in Gschem have been revised for better organization with enhanced
  options. The tools menu now includes access to Version tracking and
  alternative simulations back-ends. A toggle widget has been added to
  the Options menu to toggle a new drag-can-move preference variable in
  order to enable and disable dragging moving.

  A new File/Export main menu fly-out was added for options to export both
  embedded components and pictures. Exporting pictures utilized a new custom
  Image Chooser Dialog allowing user to select from a variety of options.
  All available image formats suppported by host libraries are supported.

  Draw-Order functions have been added to the Page menu as a fly-out using
  Scheme, allowing users to manipulate the order in which objects are drawn.

  The pop-up context menu now provides options for Copy, Move, Rotate, Mirror,
  Component Properties, Zoom to Mag, and Zoom to Selection. Additional options
  have also been add as fly-out side menus to the pop-up menu. Fly-out options
  use the X-Y location of the pointer when the context was invoked. Copy, Move,
  Rotate, and Mirror actions were reinstated and work identically to previous
  implementations.

  Additional icons are provided for "Move", Simulations, DRC, and attribute-
  reset. The default icon for "Move" was changed to have a more consistence
  theme with other editing icons.

  Menu tool-tips for File/Recent and Sessions/Restore now respond properly to
  user preferences. Tool-tips for symbol names on the Component dialog field
  shows the full path to symbol file.

  Anti-alias options were added to the Preference dialog in Gschem for the
  anti-alias method to be used by the chosen render. Modifing the anti-alias
  method longer requires a re-start for the new setting to take effect.

* Page Select Dialog has been enhanced to allow displaying file names without
  the paths, a toggle switch allows users to switch the path portion on or off
  in the tree-view. The user perference is retained between sessions.

* The appearance of Auto-Number Dialog has been improved and is more robust.
  The dialog was enhanced with frames and separate filter options for scope
  text including options for wild-cards and a list of recently used filters.
  Recent filters are not retained between sessions.

* The Line Type Dialog now includes an options combo to set the line end-type.


utils/scripts/sw2asc.awk
#!/usr/bin/mawk -f
#
# gEDA - GPL Electronic Design Automation
#
# Copyright (C) 2003-2012 Dan McMahill
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if  not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

#
# This is a script to extract out PRINT data from a SWITCAP simulation
# and write it to a format which gwave can read
#

BEGIN {
  state_wait = 0;
  state_skip_blank = 1;
  state_read_data = 2;
  state = state_wait;

  analysis_none = 0;
  analysis_SSS = 1;
  analysis_SPECTRUM = 2;
  analysis_FREQCOMP = 3;
  analysis_TRAN = 4;
  analysis = analysis_none;

  progn = "sw2asc";

  pat = "none";

  variable_offset = 0;
  nvars = 0;

  for(i=1; i<ARGC; i=i+1) {
    if(ARGV[i] ~ /^(-c)|(--copyright)$/) {
      copyright();
      exit 0;
    } else if(ARGV[i] ~ /^((-h)|(--help))$/) {
      usage();
      exit 0;
    } else if(ARGV[i] ~ /^((-v)|(--version))$/) {
      copyright();
      exit 0;
    } else if(ARGV[i] ~ /^-.*/) {
      printf("Unknown option: ""%s""\n", ARGV[i]) > "/dev/stderr";
      usage();
      close("/dev/stderr");
      exit 1;
    }
    printf("ARGV[%d] = %s\n", i, ARGV[i]);
  }
  infile = ARGV[ARGC-1];

  copyright();

  printf("Loading SWITCAP output file [%s]\n", infile);
}

/^ Sinusoidal Steady State/ {
  if( debug ) { printf("Start of SSS analysis\n"); }
  analysis = analysis_SSS;
  cnt_SSS = cnt_SSS + 1;
  fileid = ".SSS." cnt_SSS;
  pat = "^[ \t]*frequency";
}

/^ Spectral Analysis$/ {
  if( debug ) { printf("Start of SPECTRUM analysis\n"); }
  analysis = analysis_SPECTRUM;
  cnt_SPECTRUM = cnt_SPECTRUM + 1;
  fileid = ".SPECTRUM." cnt_SPECTRUM;
  pat = "^not_implemented_yet$";
  printf("Ignoring data from SPECTRUM analysis.\n");
  printf("I haven't fully worked out how to always parse this\n");
  printf("correctly.  Please write code to do this!!!\n\n");
}

/^ Spectral Analysis by Components$/ {
  if( debug ) { printf("Start of FREQCOMP analysis\n"); }
  analysis = analysis_FREQCOMP;
  cnt_FREQCOMP = cnt_FREQCOMP + 1;
  fileid = ".FREQCOMP." cnt_FREQCOMP;
  pat = "^not_implemented_yet$";
  printf("Ignoring data from FREQCOMP analysis.\n");
  printf("I haven't fully worked out how to always parse this\n");
  printf("correctly.  Please write code to do this!!!\n\n");
}

/^ Transient Analysis$/ {
  if( debug ) { printf("Start of TRAN analysis\n"); }
  analysis = analysis_TRAN;
  cnt_TRAN = cnt_TRAN + 1;
  fileid = ".TRAN." cnt_TRAN;
  pat = "^[ \t]*time";
}

/^ Run Report$/ {
  if( debug ) { printf("Start of Run Report\n"); }
  analysis = analysis_none;
  fileid = "none";
  pat = "^this_is_the_run_report$";
}

/\*SWITCAP/ {
  if( debug ) { printf("Saw *SWITCAP line\n"); }

  if( variable_offset > 0 ) {
    outf = infile fileid ".asc";
    printf("Writing data to ""%s""\n", outf);
    printf( "" ) > outf;
    for(i = 1; i <= variable_offset; i = i + 1) {
      printf("%s ", vnames[i]) >> outf;
    }
    printf("\n") >> outf;

    for(pt = 1; pt < cnt ; pt = pt + 1) {
      if( debug > 1) { printf("point #%d: ", pt); }
      for(i = 1; i <= variable_offset; i = i + 1) {
	printf("%s ", data[i ":" pt]) >> outf;
      }
      printf("\n") >> outf;
    }
    close(outf);

    variable_offset = 0;
    nvars = 0;
    for(k in data) { delete data[k]; }
    num_analysis = num_analysis + 1;

  } else {
    if( debug ) { printf("No data to output yet\n"); }
  }

  next;
}


state == state_wait {
  if( $0 ~ pat ) {
    if( debug) { printf("state_wait: Variable start line:"); }
    for(i=1; i<=NF; i=i+1) {
      if( debug ) { printf("\tVariable #%d = %s\n", i, $i); }
      vnames[i + variable_offset] = $i;
    }

    nvars = NF;

    if( debug ) { printf("state_wait:  moving to state_skip_blank\n") };
    state = state_skip_blank;
  }
  next;
}

state == state_skip_blank {
  if( $0 == "" ) {
    if( debug) { printf("state_skip_blank:  skipping blank line\n"); }
    next;
  }

  cnt = 1;
  if( debug ) { printf("state_skip_blank:  moving to state_read_data\n") };
  state = state_read_data;
}

state == state_read_data {
  if( $0 == "" ) {
    variable_offset = variable_offset + nvars;
    nvars = 0;
    if( debug ) { printf("state_read_data:  moving to state_wait\n") };
    state = state_wait;
    next;
  }

  if( debug > 1 ) { printf("state_read_data: data = %s\n", $0); }
  for(i=1; i<=NF; i=i+1) {
    gsub(/d$/, "", $i);
    data[i+variable_offset ":" cnt] = $i;
  }
  cnt = cnt + 1;
  next;
}

END {
  if( debug ) { printf("END.  Processed %d analyses.\n", num_analysis); };
}


function copyright() {
  printf("gEDA/%s version %s\n", progn, "2.0.6");
  printf("gEDA/%s comes with ABSOLUTELY NO WARRANTY; see COPYING for more details.\n", progn);
  printf("This is free software, and you are welcome to redistribute it under certain\n");
  printf("conditions; please see the COPYING file for more details.\n\n");
}

function usage() {
  printf("Usage:\n\n");
  printf("\t%s -- -c|--copyright\n", progn);
  printf("\t%s -- -h|--help\n", progn);
  printf("\t%s -- -v|--version\n", progn);
  printf("\t%s switcap.out\n", progn);
  printf("\n");
  printf("%s process the output file generated by the SWITCAP switched capacitor\n", progn);
  printf("circuit simulator.  Data generated by SWITCAP ""PRINT"" statements is\n");
  printf("extracted and written to an output file in an ASCII format which can be\n");
  printf("directly loaded by the gwave waveform viewer or more easily loaded into\n");
  printf("a tool such as scilab, octave, grace, or matlab.\n");
  printf("\n");
  printf("If multiple analyses are in the SWITCAP output file, then they are written\n");
  printf("to individual files by %s.  The output file names are derived from the input\n", progn);
  printf("file name by appending "".{SSS,TRAN}.#.asc"" to the name of the input file.\n");
  printf("SSS is used for Sinusoidal Steady State analysis output and TRAN is used for.\n");
  printf("TRANsient analysis output.  # corresponds to the analysis number (ie if you \n");
  printf("have 3 SSS analyses in your file, you will end up with .SSS.1.asc, .SSS.2.asc,\n");
  printf("and .SSS.3.asc.\n");
  printf("\n");

}

